### PR TITLE
Display dependency graph using DagreD3

### PR DESCRIPTION
**Pull request by Martchus**
* See https://progress.opensuse.org/issues/40772
* Not fancy yet but a start

### CODE REVIEW

**Comment by coolo** [comment]
/tests/2068098#settings looks like it can be better :)

**Comment by coolo** [comment]
I know this WIP, but experimented a little with it:
- it needs to be in a tab of its own. we have pretty large clusters. Appending this to the huge settings table is no longer fitting. I guess this tab could either disappear or be grayed out for dependencyless jobs
- the current parents/children can disappear then
- the NAME is too noisy, especially in large clusters
- the parallel clusters need to be a little more obvious
- job status/result needs to be part of the display
- the dots need to link back to the job

**Comment by coolo** [comment]
This might be better than what we have now, but I consider it still suboptimal (this manually drafted javascript to experiment with vis.js options

screencapture-file-home-coolo-js-lesmiserables-html-2018-10-03-11_43_59

**Comment by coolo** [comment]
For the interested: http://stephan.kulow.org/hpc.html

**Comment by coolo** [comment]
Possibly a network is just the wrong way to think about it. Random links:

https://mermaidjs.github.io/flowchart.html
https://resources.jointjs.com/demos/layout

**Comment by coolo** [comment]
Playing with mermaidjs:

screencapture-file-home-coolo-js-demo-html-2018-10-03-13_05_49

**Comment by Martchus** [comment]
> Possibly a network is just the wrong way to think about it.

I had the same thought when testing vis.js.

The mermaidjs example looks good. I'll change the PR to use that library. Shouldn't be much work if their API is similar. @coolo Do you have already code or is manually drafted?

**Comment by coolo** [comment]
Their API is completely different (and IMO sucks): http://stephan.kulow.org/mermaid.html

But if we render it like the above, I have the impression we can just as well do it on our own (and better suited to our problem).

So some more links:
https://www.codeproject.com/Articles/16192/Graphic-JavaScript-Tree-with-Layout
https://dagrejs.github.io/project/dagre-d3/latest/demo/clusters.html

**Comment by coolo** [comment]
DagreD3 is my favorite for now - e.g. it allows us to specify the parallel clusters as tables and then layout it within the graph: https://dagrejs.github.io/project/dagre-d3/latest/demo/dom.html

**Comment by Martchus** [comment]
@coolo I think I've got the algo to compute the clusters. At least the results look plausible :-)

screenshot_20181004_161458

screenshot_20181004_161531

screenshot_20181004_161515

**Comment by coolo** [comment]
And 2068098 ? :)

**Comment by Martchus** [comment]
I only have 2041334 because my database is a little bit older:

screenshot_20181004_165445

**Comment by coolo** [comment]
This leaves "the current parents/children can disappear then" - the rest works great.

**Comment by Martchus** [comment]
The reason why UI test are failing still failing:  
screenshot_20181005_131940

File: lib/OpenQA/WebAPI.pm [hunk 1]
```diff
@@ -130,6 +130,7 @@ sub startup {
     $test_r->get('/')->name('test')->to('test#show');
     $test_r->get('/ajax')->name('job_next_previous_ajax')->to('test#job_next_previous_ajax');
     $test_r->get('/modules/:moduleid/fails')->name('test_module_fails')->to('test#module_fails');
+    $test_r->get('dependencies')->name('test_dependencies')->to('test#dependencies');
```
Comment (coolo) [inline lib/OpenQA/WebAPI.pm]: for consistency either have all $test_r->get calls with / or none.

File: assets/assetpack.def [hunk 2]
```diff
@@ -100,6 +100,13 @@
 << https://raw.githubusercontent.com/twbs/bootstrap/v4.1.1/scss/_print.scss
 < https://raw.githubusercontent.com/sorich87/bootstrap-tour/6a1028fb562f9aa68c451f0901f8cfeb43cad140/build/css/bootstrap-tour.css
 
+! d3.js
```
[the remark this answers is not in the document]
Comment (Martchus) [reply assets/assetpack.def]: changed to `dagre-d3.js`

File: assets/javascripts/test_result.js [hunk 3]
```diff
@@ -335,18 +334,26 @@ function setupResult(state, jobid, status_url, details_url) {
...
     } else {
       pauseLiveView();
     }
   });
-  // start the live view if it's default link
+  // start the live view if it's default tab
   if (window.location.hash === '#live') {
     resumeLiveView();
   }
+  // setup dependency graph if it's the default tab
```
[the remark this answers is not in the document]
Comment (Martchus) [reply assets/javascripts/test_result.js]: fixed

File: assets/javascripts/test_result.js [hunk 4]
```diff
@@ -398,3 +405,135 @@ function setupResult(state, jobid, status_url, details_url) {
...
+                testResultTd.text(testResultName);
+                testResultTd.node().className = testResultId;
+
+                return table;
+            },
+            padding: 0,
+            tooltipText: node.tooltipText,
+            testResultId: testResultId,
+            testResultName: testResultName,
+            fill: "#afa",
```
[the remark this answers is not in the document]
Comment (Martchus) [reply assets/javascripts/test_result.js]: This fill is overridden in CSS anyways. I moved the other `fill` (which was actually used) to CSS.

File: lib/OpenQA/WebAPI.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI.pm]: fixed

**Pull request merged**